# Student Perfomance Indicator 
## Life Cycle of Machine Learning Project
- Understanding the problem statement
- Data collection
- Data checks to Perform 
- Exploratory Data Analysis 
- Data Pre-processing 
- Model Training
- choose best model

## 1) Problem Statement:
- This project understands how the students (test scores) is affected by other variables such as Gender, Ethnicity, Parental level of Education, Lunch and Test preparation course 

## 2) Data Collection :
- Dataset is downloaded from the kaggle
- The data consists of 8 columns and 1000 rows 

## 2.1 Import Data and Required Packages

### Importing Pandas, numpy, seaborn and Warning library.



In [2]:
import pandas as pd
import numpy as np 
import seaborn as sns 
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

In [ ]:
# Import the CSV Data as Pandas DataFrame

## Import the CSV data as Pandas Dataframe 


In [6]:
df = pd.read_csv('data/data.csv')

In [7]:
df

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75
...,...,...,...,...,...,...,...,...
995,female,group E,master's degree,standard,completed,88,99,95
996,male,group C,high school,free/reduced,none,62,55,55
997,female,group C,high school,free/reduced,completed,59,71,65
998,female,group D,some college,standard,completed,68,78,77


#### Show top 5 records (by default 5)

In [8]:
df.head()

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75


#### Shape of the dataset


In [9]:
df.shape

(1000, 8)

## 2.2 Dataset Information 
- Gender : - Sex of students -> (Male/Female)
- Race/Ethnicity : Ethnicity of the studnets :- (Group A,B,C,D)
- Parental Level of Education : parents final education -> Bachelors Degree , some college , masters degree ,  associates degree high school
- lunch :- having lunch before test ( standard or free/reduced)
- test preparation course : complete or not compelte before test 
- math score 
- reading score 
- writing score 

## 3. Data Checks to Perform 
 - handling missing values
 - duplicate values
 - data type 
 - check the number of unique values of each column 
 - check statistics of the dataset 
 - check various categories present in  the different categorical column    

 ## 3.1 Check missing Values


In [11]:
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

- There are no missing values in the dataset 

## 3.2 Check Duplicates 

In [ ]:
df.duplicated().sum()

# If Duplicate Values are present you can deal with them using 
# df.drop_duplicates()
# This drop the duplicate values present in the dataset 

0

- There are no Duplicate values in the dataset 

## 3.2 Check Datatype

In [ ]:
#check Null and Dtype 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


## 3.4 Check the number of Unique Value of Each Column 

In [15]:
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

## 3.5 Check Statistics of the Dataset

In [17]:
df.describe()

,math score,reading score,writing score
count,1000.00000,1000.000000,1000.000000
mean,66.08900,69.169000,68.054000
std,15.16308,14.600192,15.195657
min,0.00000,17.000000,10.000000
25%,57.00000,59.000000,57.750000
50%,66.00000,70.000000,69.000000
75%,77.00000,79.000000,79.000000
max,100.00000,100.000000,100.000000


## Insight

- From the above description of the numerical data, the mean scores are very close to each other, ranging between **66** and **68.05**.
- The standard deviations are also very similar, lying between **14.60** and **15.19**, indicating comparable variability across the three subjects.
- The minimum score in **Math** is **0**, whereas the minimum score in **Writing** is **10**, and in **Reading** it is **17**.

## 3.6 Exploring Data 

In [ ]:
print("Categories in Gender Variable:- ",end=" ")
print(df['gender'].unique())


# There are 2 Categories in the Gender

Categories in Gender Variable:-  ['female' 'male']


In [19]:
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

In [22]:
print("Categories in the race/ethnicity are :-",end=" ")
print(df['race/ethnicity'].unique())

Categories in the race/ethnicity are :- ['group B' 'group C' 'group A' 'group D' 'group E']


,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75
